src/extensions/sb-ncbr/partial-charges/property.ts

```diff
@@ -0,0 +1,193 @@
+
+type TypeId = number;
+type IdToCharge = Map<number, number>;
+type ChargesData = {
+    typeIdToMethod: Map<TypeId, string>;
+    typeIdToAtomIdToCharge: Map<TypeId, IdToCharge>;
+    typeIdToResidueToCharge: Map<TypeId, IdToCharge>;
+    maxAbsoluteAtomCharges: IdToCharge;
+    maxAbsoluteResidueCharges: IdToCharge;
+    maxAbsoluteAtomChargeAll: number;
+};
+type PartialCharges = PropertyWrapper<ChargesData | undefined>;
+
+let updatedParams = false;
+let PartialChargesPropertyParams = {
+    typeId: PD.Select<number>(0, [[0, '0']]),
+};
+type PartialChargesPropertyParams = typeof PartialChargesPropertyParams;
+let defaultPartialChargesPropertyParams = PD.clone(PartialChargesPropertyParams);
+
+function getParams(model: Model) {
+    // This is a hack for setting the option values for the dropdown menu
+    if (!updatedParams) {
+        const typeIdToMethod = getTypeIdToMethod(model);
+        const options = Array.from(typeIdToMethod.entries()).map(
+            ([typeId, method]) => [typeId, method] as [number, string]
+        );
+        PartialChargesPropertyParams = {
+            typeId: PD.Select<number>(1, options),
+        };
+        defaultPartialChargesPropertyParams = PD.clone(PartialChargesPropertyParams);
+        updatedParams = true;
+    }
+
+    return PartialChargesPropertyParams;
+}
+
+async function getData(model: Model): Promise<CustomProperty.Data<PartialCharges>> {
+    await Promise.resolve();
+    const info = PropertyWrapper.createInfo();
+
+    if (!isApplicable(model)) return { value: { info, data: undefined } };
+
+    const typeIdToMethod = getTypeIdToMethod(model);
+    const typeIdToAtomIdToCharge = getTypeIdToAtomIdToCharge(model);
+    const typeIdToResidueToCharge = getTypeIdToResidueIdToCharge(model, typeIdToAtomIdToCharge);
+    const maxAbsoluteAtomCharges = getMaxAbsoluteCharges(typeIdToAtomIdToCharge);
+    const maxAbsoluteResidueCharges = getMaxAbsoluteCharges(typeIdToResidueToCharge);
+    const maxAbsoluteAtomChargeAll = getMaxAbsoluteAtomChargeAll(maxAbsoluteAtomCharges);
+
+    return {
+        value: {
+            info,
+            data: {
+                typeIdToMethod,
+                typeIdToAtomIdToCharge,
+                typeIdToResidueToCharge,
+                maxAbsoluteAtomCharges,
+                maxAbsoluteResidueCharges,
+                maxAbsoluteAtomChargeAll,
+            },
+        },
+    };
+}
+
+function getTypeIdToMethod(model: Model) {
+    const typeIdToMethod: ChargesData['typeIdToMethod'] = new Map();
+
+    const sourceData = model.sourceData as MmcifFormat;
+    const rowCount = sourceData.data.frame.categories.partial_atomic_charges_meta.rowCount;
+    const typeIds = sourceData.data.frame.categories.partial_atomic_charges_meta.getField('id')?.toIntArray();
+    const methods = sourceData.data.frame.categories.partial_atomic_charges_meta.getField('method')?.toStringArray();
```

Comment:
You don't need to turn this into an array. You can access the values via `.int` or `.str` directly since you only convert this to a map immediately anyway.